src/traces/heatmap/plot.js

```diff
@@ -18,6 +18,50 @@ var LINE_SPACING = alignmentConstants.LINE_SPACING;
 
 var labelClass = 'heatmap-label';
 
+// Pixelated image rendering
+// The actual declaration is prepended with fallbacks for older browsers.
+// https://developer.mozilla.org/en-US/docs/Web/CSS/image-rendering
+// https://caniuse.com/?search=image-rendering
+var pixelatedImageCSS = [
+    'image-rendering: optimizeSpeed',
+    'image-rendering: -moz-crisp-edges',
+    'image-rendering: -o-crisp-edges',
+    'image-rendering: -webkit-optimize-contrast',
+    'image-rendering: optimize-contrast',
+    'image-rendering: crisp-edges',
+    'image-rendering: pixelated'
+];
+
+var _supportsPixelated = null;
+function supportsPixelatedImage() {
+    if(_supportsPixelated !== null) { // only run the feature detection once
+        return _supportsPixelated;
+    }
+    if(Lib.isIE()) {
+        // `-ms-interpolation-mode` works only with <img> not with SVG <image>
+        _supportsPixelated = false;
+    } else {
+        var declarations = Array.from(pixelatedImageCSS).reverse();
```

Comment:
The `pixelatedImageCSS` array is ordered so that the actual declaration `image-rendering: pixelated`, which targets modern browsers (most users) is the last one, and the oldest fallback comes first. Reversing the array is just for doing the testing on the actual declaration first, and let `declarations.some` returns early.